### PR TITLE
Fix secret scrubbing log latency

### LINKED ISSUE #5791: 🐞 secrets scrubbing implementation causes excessive log output latency
### What is the issue?

# Context

We have a dagger pipeline which, at some point, [mounts secrets](https://github.com/airbytehq/airbyte/blob/327d3c9ae8e14b28bf2bdf7528b22f86f510d629/airbyte-ci/connectors/pipelines/pipelines/actions/environments.py#L1053) to a container which is then runs [integration tests](https://github.com/airbytehq/airbyte/blob/327d3c9ae8e14b28bf2bdf7528b22f86f510d629/airbyte-ci/connectors/pipelines/pipelines/gradle.py#L157) by with_exec'ing a task in gradle.

We are just as bad computer programmers as anyone and sometimes we introduce regressions which cause our integration tests to hang. In our particular case, the bug we introduced was that we tried to connect a jdbc driver to a nonexistent host and we'd hit some absurdly large timeout in our code.

What's unexpected, and why I'm filing an issue right now, is that dagger would be completely silent about the output of this `with_exec` step until it eventually failed. 

# The problem

I eventually found that dagger's secrets scrubbing implementation is to blame. In our case, we mount about 20 secrets files, which while a lot, doesn't seem that crazy. I read dagger's code [1] and it appears that if there's a nonzero number of secrets mounted (env vars or files) the output gets piped through `replace.Chain` which chains a bunch of `Transformer` instances from go's transform module [2]. There's one instance per secret and each appears to wrap a reader with a hardcoded buffer of 4kB [3].

In our case, our gradle process was simply not chatty enough to overcome this buffering and so it seemed like it didn't do anything. We're on an old version of dagger but the problem seems still present in the `main` branch from what I can tell.

# My expectations as a user

The need to scrub secrets is not in question. However, `transform` appears to be written in a way that biases throughput over latency, when in our case exactly the opposite is desirable: when scrubbing secrets my expectation is that the output becomes visible to me _as soon as possible_. I don't want to wait for some arbitrary unconfigurable buffer to fill up and I certainly don't want that behaviour to be `O(n)` to the number of secrets mounted.

I say as soon as possible but that's not strictly speaking true, it's not like each rune has to pass through as soon as it's determined that it's not part of a prefix of one of the secrets. This being textual output, I'd be fine if it was each _line_ that gets passed through ASAP. I'm guessing this would also be undoubtedly less complicated to implement on your end and I'm also guessing that most secrets are not multiline (or if they are, it's only a handful of lines). 

tl;dr: no need to go crazy here, but I'll be happy with reasonable and predictable behaviour.

# Repro

I didn't look into how dagger is tested but the desired behaviour of the return value of `NewSecretScrubReader` seems adequately unit-testable from my naive standpoint:
1. mock mounting 1, 10 and 100 secrets in each test case;
2. kick off a goroutine with an `io.Reader` which mocks a `with_exec`'ed process' `stdout`;
3. have that `io.Reader` serve a fixed number of bytes and then block on a channel, then serve a bunch more bytes, then block again;
4. concurrently, expect the `io.Reader` returned by `NewSecretScrubReader` to read an expected subset of those bytes, then unblock the channel, then read another expected subset of bytes, then unblock again.

# Notes

Dear dagger maintainers: I hope you're not finding me overprescriptive here or feel that I'm babying you. Far from it! I'm being deliberately wordy because I'm reasonably hopeful that this issue is quite cheap to fix so I'm provide as much info as I can to that aim. On our end, it sure caused a lot of head-scratching. 

This is only an issue in an interactive setting: developing locally, or looking at dagger.cloud, which is probably why it went unnoticed (I didn't find any similar issue, though I only did a quick search).

# References

[1] https://github.com/dagger/dagger/blob/v0.6.4/cmd/shim/main.go#L294
[2] https://golang.org/x/text/transform
[3] https://cs.opensource.google/go/x/text/+/refs/tags/v0.13.0:transform/transform.go;l=130

### Log output

_No response_

### Steps to reproduce

_No response_

### SDK version

0.6.4

### OS version

irrelevant, I believe, but macOS and some unremarkable linux distro

### PR DESCRIPTION
**Pull request by jedevc**
Fix #5791

This was a fun exercise in processing streams in go, and an absolutely massive nerd-snipe :cry: 

Essentially, we need a custom transformer to handle *precisely* matching Reads on the underlying source with the output - we shouldn't hold output any longer than is absolutely neccessary (which was the issue with the previous implementation).

To be able to do this at all in any reasonable way, we need a trie, and handle *all* the secrets at once, instead of doing multiple passes. Multiple passes won't work, since it's possible to accidentally trim too much at each step, which would be very sad.

> e.g. imagine secrets (aaa, bbb, ccc, etc), and an input (cba)
>
> In removing secret aaa, we would trim to cb, then we'd remove bbb to trim to c, then finally trim to nothing. However, this is overly enthusiastic, we could easily just trim to cb, if we knew about all the secrets at once.

So, we need a trie, and we need a custom implementation of one. This is because *no off the shelf implementation* seems to allow traversing the trie state-by-state. Thankfully, it's a pretty short implementation to implement one from scratch, and not too much harder to turn it into a radix tree (which lets us use quite a lot less memory).

With our trie, we can implement our custom transformer, which is *an utter pain*. Honestly, the comments should explain all the fun edge cases it's possible to hit. There's a lot of tests added as well, each of which was a real horrible thing I hit while implementing it.

I played around a bit with benchmarking, but ugh, it's a *tiny* bit slower than the original implementation (maybe by like ~25%?). It's not huge, but the latency problem is **actually solved**. Some potential things I did look into and gave up on:
- Only copy into dstBuf when dst is full (requires tons of extra conditionals, so slows everything down).
- Avoid copies at all costs by having "virtual buffer pointers" into src, that indicate future data to copy (not only is this *slower*, the logic becomes truly incomprehensible).
- Playing with off-the-shelf radix tree implementations, but they're so inconvenient to use for this specific use case, it'd be way more trouble than it'd be worth.

Any ideas welcome, but honestly, I've looked at enough flamegraphs today.

### CODE REVIEW

File: internal/mage/engine.go [hunk 1]
```diff
@@ -324,6 +325,37 @@ func (t Engine) Dev(ctx context.Context) error {
 	return nil
 }
 
+func (t Engine) TestShim(ctx context.Context) error {
```
Comment (TomChv) [inline internal/mage/engine.go:328]: Could you remove that before merge, it was only for testing purpose

**Review by TomChv — APPROVED** [review]
Nice implementation, it looks good and tests passes, I let @sipsma, @vito or @helderco double check but lgtm

File: internal/mage/engine.go (hunk 1 above)
Comment (jedevc) [reply internal/mage/engine.go]: Done, thanks :tada:

**Review by vito — APPROVED** [review]
Super cool to see the "final form" of this feature after being really hesitant to even support it in Concourse due to the complexity involved. ("Simply don't log secrets," I said!)

Latency wise this definitely passes the eye test. After Daggerizing the Daggerverse yesterday I noticed logs for `postgres` would occasionally "hitch" - I see no such thing after this change. Nicely done!

Also thanks for all the commentary in the code. :+1:

File: cmd/shim/secret_scrub_test.go [hunk 2]
```diff
@@ -241,3 +247,329 @@ func TestScrubSecretWrite(t *testing.T) {
...
+			require.Equal(t, strings.Repeat("a", 4096), string(buf[:n]))
+			n, err = r.Read(buf)
+			require.NoError(t, err)
+			require.Equal(t, strings.Repeat("a", 4096), string(buf[:n]))
+			n, err = r.Read(buf)
+			require.ErrorIs(t, err, io.EOF)
+			require.Equal(t, strings.Repeat("a", 1807), string(buf[:n]))
+			done.Add(1)
+		}()
+		require.Eventually(t, func() bool {
```
[the remark this answers is not in the document]
Comment (jedevc) [reply cmd/shim/secret_scrub_test.go]: Yeah 100% agreed - have updated to do this instead.

**Comment by vito** [comment]
Will babysit this to get it merged, it was only failing on a security scan which I'm pretty sure is fixed in `main`.
